### ISSUE TITLE
Types of `add` and `multiply` are too loose

### ISSUE THREAD

**Issue by lexi-lambda** [mniip/finite-typelits#13]
The fact that `Finite n` represents a natural number in the *open* range [0, *n*), equivalent to the closed range [0, *n* − 1] rather than the closed range [0, *n*], seems to be a consistent source of imprecisions in this library’s types. For example, the current type of `add` is

```haskell
add :: Finite n -> Finite m -> Finite (n + m)
```

but this is too loose, since it implies the result of summing a `Finite n` value with a `Finite m` value will produce a value in the range [0, *m* + *n*). However, the maximum value is actually (*n* − 1) + (*m* − 1), or *n* + *m* − 2, which means the type above includes an off-by-one error. The actual type should be

```haskell
add :: Finite n -> Finite m -> Finite (n + m - 1)
```

to reflect the actual upper bound.

The situation for `multiply` is similar, but the differences between the operations make the inaccuracy more drastic than a simple off-by-one error. The current type of `multiply` is

```haskell
multiply :: Finite n -> Finite m -> Finite (n * m)
```

which implies the range of `Finite n` multiplied by `Finite m` is [0, *n* × *m*), which would be equivalent to the closed range [0, (*n* × *m*) − 1]. This is not true, however, since the same reasoning used above reveals that the possible range is actually [0, ((*n* − 1) × (*m* − 1)) + 1). This implies that the correct type for `multiply` should be

```haskell
multiply :: Finite n -> Finite m -> Finite (((n - 1) * (m - 1)) + 1)
```

messy as that may be.

**Comment by mniip** [comment]
Good catch. I wonder if there's anything that can be done about it that's less messy than the `-1` and `+1` all over the place. One point of trouble is that
```hs
add @0 @0 :: Finite 0 -> Finite 0 -> Finite (0 + 0 - 1)
```
which happens to be even worse for `multiply`.

**Comment by lexi-lambda** [comment]
Why is the issue with `add @0 @0` actually a problem? `Finite 0` is uninhabited, so `add @0` is `absurd`.

**Comment by Blaisorblade** [comment]
(Sorry to butt in from other dependently-typed languages and Twitter).

I'd be tempted to write 
```haskell
add:: Finite (S n) -> Finite (S m) -> Finite (S (m + n))
```

That can be even more accurate, since we're using saturating subtraction `0 - 1 = 0`. But to use this, you need to prove that your input natural is non-zero first!

Still I wonder about the overhead, also because even the Agda std. library uses the looser type (https://github.com/agda/agda-stdlib/blob/983e1487e67d02ff9f3ee79c74e4f959e98aa799/src/Data/Fin.agda#L150), maybe because nobody noticed?

## Digression from use case

More interestingly, I wonder what it'd take to typecheck something like the following (liberally reverse-engineered pseudocode from @lexi-lambda's https://twitter.com/lexi_lambda/status/1029918793034805248). Most types I can conceive for the involved operators require a spurious case distinction on `n` and `o` to even hope for this to typecheck.

```haskell
chunks :: (KnownNat n, KnownNat o, KnownNat p, n ~ o * p) => Sized.Vector n a -> Sized.Vector o (Sized.Vector p a)
chunks xs = [ [ Sized.index xs (i * p + j) | j <- until p ] | i <- until o]]
-- This uses `i * p`, not `i * (maxBound :: Finite p) = i * (p - 1)`.

-- Proof obligation: `0 <= i < o and 0 <= j < p implies i * p + j <= o * p`.
-- My best proof requires using integer subtraction and going through negative numbers: We have i <= o - 1, so i * p <= (o - 1) * p = n - p, and i * p + j <= n - p + j = n - (p - j) < n.

-- returns [0..i[, *not* [0..i-1]
until :: i -> [Finite i] 
until 0 = []
until (S i) = until i ++ [finite i]
```

But I guess this is out-of-scope; a dependently-typed programmer would rather write something else (or give up altogether?), while a Coq user might eschew dependent types and verify correctness externally.

EDIT: dunno how powerful inequality constraints are in GHC, but they seem more promising?

**Comment by mniip** [comment]
As a workaround, you can use ``(+) `on` weakenN``.

**Comment by mniip** [comment]
@Blaisorblade This library has a `separateProduct` function.

**Comment by jonathanknowles** [comment]
Hi @mniip!

Just wondering if you'd be open to a PR that addresses this issue?

For the problem you mentioned above:
```hs
add @0 @0 :: Finite 0 -> Finite 0 -> Finite (0 + 0 - 1)
```

Since addition is presumably not defined in cases where `n == 0` or `m == 0`, perhaps we could simply constrain `n` and `m` to be non-zero:
```hs
add :: forall n m. (1 <= n, 1 <= m)
    => Finite n -> Finite m -> Finite (n + m - 1)
```
This way, the result is now a `Finite r` such that `1 <= r`.

Similarly for `multiply`, we could write:
```hs
multiply
  :: forall n m. (1 <= n, 1 <= m)
  => Finite n -> Finite m -> Finite (((n - 1) GHC.TypeLits.* (m - 1)) + 1)
```
The result is again a `Finite r` such that `1 <= r`. Justification:
```hs
    1 <=  n      , 1 <= m
=>  0 <=  n - 1  , 0 <= m - 1
=>  0 <= (n - 1) *     (m - 1)
=>  1 <= (n - 1) *     (m - 1) + 1
```

I've prototyped this here, for reference:
- [📌 `0ec95be4855b0b2f` Tighten type of `add`.](https://github.com/jonathanknowles/finite-typelits/commit/0ec95be4855b0b2fd32b708eba9c9ba04dce0f30)
- [📌 `4b2ad176d3a8840c` Tighten type of `multiply`.](https://github.com/jonathanknowles/finite-typelits/commit/4b2ad176d3a8840cddce9fa54b39a02cae136db0)

I'd be happy to turn this into a PR if you think this approach could work. Or I'd be glad to hear any alternative ideas you might have.
